### ISSUE TITLE
Specify source facility and jurisdiction in scraped data

### ISSUE THREAD

**Issue by rasmi** [Recidiviz/recidiviz-data#386]
In order to run queries such as
```
SELECT COUNT(DISTINCT(person_id))
FROM person
GROUP BY jail
```
We need to be able to disambiguate which physical location (jail/facility) and which authority/jurisdiction the data is tied to. Ideally, each record (at a booking level?) could be tied to some detailed facility code. We would need to add this information with minimal disruption to the scraper writers. 

We would also need to be able to handle:
* Websites that serve multiple counties or facilities and don't distinguish between them.
* Websites where one facility is holding people for one or more jurisdictions (some notion of physical location/facility vs. authority/custody).
* Facility data (capacity, for example) changing over time, and understanding how many people were in a facility at some point in time in the context of those historical properties.
* Adding new facilities over time without changing IDs for other facilities (IDs are globally unique and deterministically generated).

Anything else?

**Comment by rasmi** [comment]
A first pass at a schema that encapsulates some of the ideas the Vera team has come up with:

There is a list of "Authorities" (also/formerly known as "Jurisdictions") that have one or more "Facilities" (physical locations, a collection of one or more buildings, where someone is being held). 

An `authority` table that has the following columns:
* authority_id (primary key)
* authority_name (string)
* authority_type (enum)
* misc. other properties (please specify more if I've missed anything!)

A `facility` table that has the following columns:
* `facility_id` (unique alphanumeric key)
* `authority_id` (foreign key)
* `facility_type` (enum)
* `location` (?)
* `location_state` (int state FIPS code)
* `location_county` (int state+county FIPS code)
* `date_opened` (datetime)
* `date_closed` (datetime)
* Various physical properties: `capacity`, `num_buildings`, `num_cells`, `num_beds`, etc. Some of the examples I've listed are building-level and it's not clear how to handle individual buildings yet.

In this structure, the facility is the most granular level of detail and each facility can only belong to one authority. This doesn't account for individual buildings in a facility, but I think that's something we can ignore for the purposes of this project, or add in later (where building has a foreign key to facility) if we keep our definitions of facility well enough.

Additionally, at a `booking` level:

`booking`:
* `facility` (foreign key)
* `custody_authority`(foreign key) -- equivalent to `booking.facility.authority_id`, but we may not have `facility` values for all `booking`s.
* `hold` (effectively the authority, but not a key, just a string)

Thoughts?

**Comment by jessex** [comment]
Great start, @rasmi! Some initial questions after reading through:
* Should an authority have a distinct region, or maybe a set of regions? I'm thinking you may want to calculate metrics by authority for all authorities within a given region, to compare how different administrative bodies are performing relative to one another.
* This may be covered by authority_type but I think it would be useful to know the _scope_ of the authority: jails, prisons, community supervision, some combination of these?
* I think it's likely that virtually all facilities have only a single authority we care to track, but I think facilities change authority from time to time. Do you envision supporting history tables for these new entities? If not, how would we model change over time, be it administrative (new authority takes over), physical (expansion of the facility), or otherwise?
* I agree that getting down to the building level is out of scope for this work but a potentially useful enhancement in the more distant future. But part of the reason I think of it as out of scope is because of the challenge of coming by all of this information. And I do wonder how much of the facility-level information we'll be able to come by, anyway. @ohinds once relayed that he wasn't aware of any publicly available data set on jails or prisons, for example. The big question: How do we gather this information and keep it up-to-date? Do you envision we'll be able to get much of this from scraping? That seems unlikely to me, but you've probably scraped as much as I have at this point.
* Are facilities for community supervision, such as halfway houses, included as facilities?
* Does `booking.hold` map (effectively) to `authority.authority_name`?
* One of the things you call out as a product requirement: "Websites where one facility is holding people for one or more jurisdictions (some notion of physical location/facility vs. authority/custody)." I think an edge case in here is a facility nominally under control of one authority holding a person on behalf of a different authority entirely, even if temporarily. How is that handled in this schema, if at all? With a booking pointing to the authority a person is being held under and that same booking pointing to the facility the person is being held within that itself points to a different authority?

**Comment by rasmi** [comment]
Thanks for the comments Josh! Some clarification!

* I think the sort of analysis you're describing could be done on the bookings, by `booking.hold` and `booking.custody_authority` (where `booking.facility`, if available, will give the location granularity).
* I don't know how detailed Vera's authority taxonomy will be, but my understanding was that it is something along the lines of municipal/state/federal etc (the extent of the legal domain/jurisdiction), and there exists a large taxonomy of facility types (prison, jail, mental health institution, etc).
* Yes, these tables should be historical like all others in the database. I haven't ironed out the details with @zdg2102, but hopefully the query style becomes clear once the implementation is done. The specific scenario we considered was marking people as "`inferred_released`" when their facility is closed, even if they're just being moved to another facility. 
* The Vera folks have a better idea on building info. I think it would be useful for a select few facilities that probably have the info available, in which case it would be manually collected and entered into the database using some other visual database editor type application.
* They are in Vera's taxonomy! It's under development so you should reach out to @ohinds if you're interested.
* `booking.hold` is effectively `authority.authority_name`, but it's important to distinguish that `booking.hold` is a free-form string field that doesn't really map to anything. Ideally we could have a `booking.hold_authority` that is always filled, but the burden on scraper-writers to map each booking to an `authority` may be too large. @ohinds argued that analysis of such a field is rare enough that it's fine to leave `booking.hold` as free-form text and do some custom computation after the fact. I think it may be worth having a `booking.hold_authority` foreign key field added just in case it's easy to map to the holding authority in some cases, in addition to the free-form `booking.hold`. The same argument can be made for having a free-text `booking.facility_name` rather than just `booking.facility`.

What do you think?

**Comment by zdg2102** [comment]
Yes, please, I'd like to keep the model simple and just have every table in the database that corresponds to an external real-world entity have a historical table and follow the same historical pattern as the other tables.

Also @rasmi two questions:

1) How do you envision scraper writers indicating which authority or facility their scraped people are associated with? We definitely don't want them hardcoding database primary keys, so would we want them to have globally unique natural-language-ish string identifiers (e.g. "NewYorkState")?

2) Why do you have facility_id called out as alphanumeric, rather than a standard integer primary key?

**Comment by jessex** [comment]
>I think it may be worth having a booking.hold_authority foreign key field added just in case it's easy to map to the holding authority in some cases, in addition to the free-form booking.hold. The same argument can be made for having a free-text booking.facility_name rather than just booking.facility.

This sounds appropriate to me at first glance. I think we should treat booking.hold and booking.facility consistently, in terms of available fields.

I also second @zdg2102's question 1, and it relates to my 4th bullet point. You responded to that bullet point that we'd want this data manually entered, but various aspects of a facility change semi-regularly, and political/bureaucratic changes in authority are not totally rate, either. So on one hand there's the challenge of keeping this data relevant and fresh so that it's useful. And on the other hand, there's the challenge of ensuring scraper writers have an easy means to make use of it in the (majority?) case where it cannot be automatically extracted from the scraped site.

**Comment by terinpw** [comment]
Chatted w @rasmi. @zacharylawrence and I added jurisdiction_ids to all people in our scraped data based on `jid.csv` generated by @eitalse. 

In the above design there was:
- Custody authority (free text)
- Custody foreign key
- Hold authority (free text)
- Hold foreign key

In our current schema we support everything in the above list except for "Hold foreign key":
- Custody authority -> `person.region`
- Custody foreign key -> `person.jursidiction_id`
- Hold authority -> `hold.jurisdiction_name`

If someone decides that we want to add an hold foreign key, this bug should be reopened.
